### PR TITLE
test: add mid price fn to ExchangePriceCacheInner

### PR DESCRIPTION
**Pull request by tvpeter**
- add mid-price for bid and ask price
- add unit test for bid, ask, and mid-price

### CODE REVIEW

File: price-server/src/server/mod.rs [hunk 1]
```diff
@@ -135,9 +135,18 @@ impl Price for PriceService {
...
+        request: Request<GetCentsPerSatsExchangeMidRateRequest>,
     ) -> Result<Response<GetCentsPerSatsExchangeMidRateResponse>, Status> {
-        Err(Status::unimplemented(""))
+        let req = request.into_inner();
+
+        let mid_amount_in_cents = self
+            .app
+            .get_cents_per_sats_exchange_mid_rate(Sats::from_major(req.amount_in_satoshis))
+            .await?;
+        let ratio_in_cents_per_satoshis = f64::from_bits(mid_amount_in_cents);
```
Comment (bodymindarts) [inline price-server/src/server/mod.rs:146]: this line looks dangerous. I would prefer a more explicit conversion (that is unit tested).

File: price-server/src/exchange_price_cache.rs [hunk 2]
```diff
@@ -95,4 +100,61 @@ impl ExchangePriceCacheInner {
         }
         Err(ExchangePriceCacheError::NoPriceAvailable)
     }
+
+    fn mid_price_of_one_sat(&self) -> Result<UsdCents, ExchangePriceCacheError> {
```
Comment (bodymindarts) [inline price-server/src/exchange_price_cache.rs:104]: What you've done here is correct - following the existing pattern...

however I think we should cleanup the `ExchangePriceCache`.

It should just expose the `BtcSatTick` directly:
```
impl ExchangePriceCache {
  pub fn get_latest_tick()
}
```

Then in the us case you can call:
```
cache.get_latest_tick().mid_price_of_one_sat()
```

then we don't have to duplicate all the use case functions 3 times.

File: price-server/src/app/mod.rs [hunk 3]
```diff
@@ -8,7 +8,7 @@ pub use error::*;
 use shared::{currency::*, payload::OkexBtcUsdSwapPricePayload, pubsub::*};
 
 pub struct PriceApp {
-    price_cache: ExchangePriceCache,
+    pub price_cache: ExchangePriceCache,
```
Comment (bodymindarts) [inline price-server/src/app/mod.rs:11]: Why is this made public?

File: price-server/src/exchange_price_cache.rs [hunk 4]
```diff
@@ -30,23 +29,22 @@ impl ExchangePriceCache {
...
     timestamp: TimeStamp,
-    ask_price_of_one_sat: UsdCents,
-    bid_price_of_one_sat: UsdCents,
+    pub ask_price_of_one_sat: UsdCents,
+    pub bid_price_of_one_sat: UsdCents,
+}
+
+impl BtcSatTick {
+    pub fn mid_price_of_one_sat(&self) -> UsdCents {
+        (self.bid_price_of_one_sat.clone() + self.ask_price_of_one_sat.clone()) / 2
```
Comment (bodymindarts) [inline price-server/src/exchange_price_cache.rs:46]: Too much `.clone()`ing. You need to `impl Add` on the pointer type `&UsdCents`.
Something like:
```
        impl std::ops::Add<&$name> for &$name {
            type Output = $name;

            fn add(self, rhs: &$name) -> Self::Output {
                Self {
                    inner: self.inner + res
                }
            }
        }

```

File: shared/src/currency/mod.rs [hunk 5]
```diff
@@ -50,6 +50,26 @@ macro_rules! currency {
             }
         }
 
+        impl std::ops::Add<$name> for $name {
```
Comment (bodymindarts) [inline shared/src/currency/mod.rs:53]: Change this to be implemented on the pointer type `&$name` or add it.

File: shared/src/payload/primitives.rs [hunk 6]
```diff
@@ -38,6 +46,19 @@ impl PriceRatioRaw {
     }
 }
 
+pub struct ConvertU64ToF64 {
```
Comment (bodymindarts) [inline shared/src/payload/primitives.rs:49]: This is a pointless struct. It doesn't add anything you can't do with stdlib - delete it.

File: price-server/src/server/mod.rs [hunk 7]
```diff
@@ -143,7 +143,12 @@ impl Price for PriceService {
             .app
             .get_cents_per_sats_exchange_mid_rate(Sats::from_major(req.amount_in_satoshis))
             .await?;
-        let ratio_in_cents_per_satoshis = f64::from_bits(mid_amount_in_cents);
+        let ratio_in_cents_per_satoshis = ConvertU64ToF64 {
```
Comment (bodymindarts) [inline price-server/src/server/mod.rs:146]: This struct for converting u64 -> f64 is not appropriate. Use stdlib or a cast.

File: shared/src/currency/mod.rs [hunk 8]
```diff
@@ -59,7 +79,6 @@ macro_rules! currency {
         }
     };
 }
-
```
Comment (bodymindarts) [inline shared/src/currency/mod.rs:62]: why delete this line break?

File: price-server/src/server/mod.rs (hunk 1 above)
Comment (tvpeter) [reply price-server/src/server/mod.rs]: extracted to a new fn and unit tested

File: price-server/src/server/mod.rs [hunk 9]
```diff
@@ -135,9 +135,23 @@ impl Price for PriceService {
...
         &self,
-        _request: Request<GetCentsPerSatsExchangeMidRateRequest>,
+        request: Request<GetCentsPerSatsExchangeMidRateRequest>,
     ) -> Result<Response<GetCentsPerSatsExchangeMidRateResponse>, Status> {
-        Err(Status::unimplemented(""))
+        let req = request.into_inner();
+
+        let mid_amount_in_cents = self
+            .app
+            .get_cents_per_sats_exchange_mid_rate(Sats::from_major(req.amount_in_satoshis))
```
Comment (bodymindarts) [inline price-server/src/server/mod.rs:144]: I'm confused about this... the proto definition says it doesn't have a field:
```
message GetCentsPerSatsExchangeMidRateRequest {}
message GetCentsPerSatsExchangeMidRateResponse {
  double ratio_in_cents_per_satoshis = 1;
}
```
How does this compile?

File: price-server/src/server/mod.rs (hunk 7 above)
Comment (bodymindarts) [reply price-server/src/server/mod.rs:146]: It should simply return the price of 1 sat as double. That implies a conversion from `impl From<UsdCent> for f64`.

**Review by bodymindarts — CHANGES_REQUESTED** [review]

File: price-server/src/exchange_price_cache.rs (hunk 4 above)
Comment (tvpeter) [reply price-server/src/exchange_price_cache.rs:46]: I have a blocker implementing `Add` for the pointer type `&$name`. I will need assistance on this. Thank you.

File: shared/src/payload/primitives.rs [hunk 10]
```diff
@@ -9,6 +11,12 @@ crate::string_wrapper! { CurrencyRaw }
 
 const PRICE_IN_CENTS_PRECISION: u32 = 12;
 
+#[derive(Error, Debug)]
+pub enum ConvertU64ToF64Error {
```
Comment (bodymindarts) [inline shared/src/payload/primitives.rs:15]: think this error can be deleted

File: CHANGELOG.md [hunk 11]
```diff
@@ -76,7 +76,7 @@
 - More price-server scafolding
 - Return pined stream from subscribe
 - Clippy
-- Commit Cargo.lock as we are buiding a binary
+- Commit Cargo.lock as we are building a binary
```
Comment (bodymindarts) [inline CHANGELOG.md:79]: remove this correction - I've taken `CHANGELOG.md` out of the spell checking.

**Review by bodymindarts — APPROVED** [review]

**Pull request closed without merging**